src/fcc_bridge_node_ros.cpp

```diff
@@ -0,0 +1,225 @@
+                this->get_name());
+}
+
+void FCCBridgeNode::setup_ros() {
+    RCLCPP_DEBUG(this->get_logger(), "Setting up ROS");
+    // Check if the node is in the correct state
+    if (this->internal_state != INTERNAL_STATE::STARTING_UP) {
+        RCLCPP_ERROR(this->get_logger(),
+                     "Repeated try to setup ros components");
+        this->internal_state = INTERNAL_STATE::ERROR;
+        return;
+    }
+
+    // Setup publisher
+
+    // Create GPSInfo & Position publisher
+    this->gps_position_publisher =
+        this->create_publisher<interfaces::msg::GPSPosition>("uav_gps_position",
+                                                             1);
+    // Create FlightState publisher
+    this->flight_state_publisher =
+        this->create_publisher<interfaces::msg::FlightState>("uav_flight_state",
+                                                             1);
+
+    // Create BatteryState publisher
+    this->battery_state_publisher =
+        this->create_publisher<interfaces::msg::BatteryState>(
+            "uav_battery_state", 1);
+
+    // Create RCState publisher
+    this->rc_state_publisher =
+        this->create_publisher<interfaces::msg::RCState>("uav_rc_state", 1);
+
+    // Setup subscriber
+
+    // Create subscription options for heartbeat to only receive mission control
+    // heartbeats
+    rclcpp::SubscriptionOptions subscription_options;
+    subscription_options.content_filter_options.filter_expression =
+        "sender_id = 'mission_control'";
+
+    // Create Heartbeat subscription
+    this->mission_control_heartbeat_subscriber =
+        this->create_subscription<interfaces::msg::Heartbeat>(
+            "heartbeat", 1,
+            std::bind(&FCCBridgeNode::mission_control_heartbeat_subscriber_cb,
+                      this, std::placeholders::_1),
+            subscription_options);
+
+    // Ensure that the filter is enabled. (Some DDS versions do not support it)
+    if (!this->mission_control_heartbeat_subscriber->is_cft_enabled()) {
+        RCLCPP_FATAL(this->get_logger(), "Content filtering is not enabled!");
+        this->internal_state = INTERNAL_STATE::ERROR;
+        return;
+    }
+
+    // Setup 5Hz timer to get telemetry from the FCC
+    this->fcc_telemetry_timer_5hz = this->create_wall_timer(
+        std::chrono::milliseconds{200},
+        std::bind(&FCCBridgeNode::fcc_telemetry_timer_5hz_cb, this));
+
+    // Setup 10Hz timer to get telemetry from the FCC
+    this->fcc_telemetry_timer_5hz = this->create_wall_timer(
+        std::chrono::milliseconds{100},
+        std::bind(&FCCBridgeNode::fcc_telemetry_timer_10hz_cb, this));
+}
+
+void FCCBridgeNode::mission_control_heartbeat_subscriber_cb(
+    const interfaces::msg::Heartbeat &msg) {
+    RCLCPP_DEBUG(this->get_logger(), "Received heartbeat from mission control");
+
+    // Check if the heartbeat tick has increased since last time
+    if (msg.tick <= this->last_mission_control_heartbeat.tick) {

```

Comment:
Please make sure to use the same tick definition I do:
*The tick must be greater than the tick before or 0, but never the same as the tick before*

Reply:
This should cover the inverse case.
Note that the if is entered if the current msg tick is the same or less than the last one.
The initial check with 0 will be done by initializing the cached heartbeat in the constructor.

Reply:
Initialization added in 0c3c010

Reply:
I had that concern in the beginning as well, but assuming a 2Hz heartbeat the tick will only overflow after 2^31 seconds which is roughly equal to 600.000 hours or 70 years.
If a node starts publishing way to quickly I would see that as an anomaly and terminate nevertheless on the overflow.
Thats at least my reasoning.

Reply:
Yes, but that's the definition and check I'm sticking to.
If we change the heartbeat rate, your function might fail.

Reply:
Ok, if we assume that an overflow is ok I will do so as well.

Reply:
Fixed in 829f417